File: frontend/src/api/camera.ts
import { request } from './client'
import type { Camera, CameraCreate, CameraListResponse, CameraUpdate } from '../types/camera'

export interface ListParams {
  page?: number
  size?: number
  ip_address?: string
  login?: string
  type?: string
}

export function listCameras(params: ListParams = {}): Promise<CameraListResponse> {
  const q = new URLSearchParams()
  if (params.page) q.set('page', String(params.page))
  if (params.size) q.set('size', String(params.size))
  if (params.ip_address) q.set('ip_address', params.ip_address)
  if (params.login) q.set('login', params.login)
  if (params.type) q.set('type', params.type)
  return request(`/cameras/?${q}`)
}

export function createCamera(data: CameraCreate): Promise<Camera> {
  return request('/cameras/', { method: 'POST', body: JSON.stringify(data) })
}

export function updateCamera(id: number, data: CameraUpdate): Promise<Camera> {
  return request(`/cameras/${id}`, { method: 'PATCH', body: JSON.stringify(data) })
}

export function deleteCamera(id: number): Promise<void> {
  return request(`/cameras/${id}`, { method: 'DELETE' })
}

File: backend/app/routers/camera.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.deps import get_session, require_permission
from app.exceptions.camera import (
    CameraAlreadyExistsException,
    CameraNotFoundException,
    DatabaseException,
)
from app.repositories.camera import CameraRepository
from app.schemas.camera import (
    CameraCreate,
    CameraListResponse,
    CameraRead,
    CameraSearchParams,
    CameraUpdate,
    PaginationParams,
)
from app.services.camera import CameraService

router = APIRouter(prefix="/cameras", tags=["cameras"])


def get_camera_service(
    session: AsyncSession = Depends(get_session),
) -> CameraService:
    return CameraService(CameraRepository(session))


def _handle(exc: CameraNotFoundException | CameraAlreadyExistsException | DatabaseException) -> None:
    raise HTTPException(status_code=exc.status_code, detail=exc.detail)


@router.post(
    "/",
    response_model=CameraRead,
    status_code=201,
    dependencies=[Depends(require_permission("cameras:write"))],
)
async def create_camera(
    data: CameraCreate,
    service: CameraService = Depends(get_camera_service),
) -> CameraRead:
    try:
        return await service.create(data)
    except (CameraAlreadyExistsException, DatabaseException) as e:
        _handle(e)


@router.get(
    "/",
    response_model=CameraListResponse,
    dependencies=[Depends(require_permission("cameras:read"))],
)
async def list_cameras(
    page: int = 1,
    size: int = 10,
    ip_address: str | None = None,
    login: str | None = None,
    type: str | None = None,
    service: CameraService = Depends(get_camera_service),
) -> CameraListResponse:
    pagination = PaginationParams(page=page, size=size)
    search = CameraSearchParams(ip_address=ip_address, login=login, type=type)
    return await service.list(pagination, search)


@router.get(
    "/{camera_id}",
    response_model=CameraRead,
    dependencies=[Depends(require_permission("cameras:read"))],
)
async def get_camera(
    camera_id: int,
    service: CameraService = Depends(get_camera_service),
) -> CameraRead:
    try:
        return await service.get(camera_id)
    except CameraNotFoundException as e:
        _handle(e)


@router.patch(
    "/{camera_id}",
    response_model=CameraRead,
    dependencies=[Depends(require_permission("cameras:update"))],
)
async def update_camera(
    camera_id: int,
    data: CameraUpdate,
    service: CameraService = Depends(get_camera_service),
) -> CameraRead:
    try:
        return await service.update(camera_id, data)
    except (CameraNotFoundException, DatabaseException) as e:
        _handle(e)


@router.delete(
    "/{camera_id}",
    status_code=204,
    dependencies=[Depends(require_permission("cameras:delete"))],
)
async def delete_camera(
    camera_id: int,
    service: CameraService = Depends(get_camera_service),
) -> None:
    try:
        await service.delete(camera_id)
    except (CameraNotFoundException, DatabaseException) as e:
        _handle(e)
